lib/Sema/SemaBounds.cpp

```diff
@@ -613,14 +614,21 @@ namespace {
             llvm_unreachable("unexpected cast failure");
             return CreateBoundsNone();
           }
+
           Expr *subExpr = CE->getSubExpr();
           BoundsExpr *Bounds = CE->getBoundsExpr();
-          QualType QT = subExpr->getType();
-	  Expr *Base = subExpr;
-          if (subExpr->getStmtClass() != clang::Stmt::CallExprClass)
-	    Base=CreateImplicitCast(QT, CastKind::CK_LValueToRValue, subExpr);
+
```

Comment:
I wanted to follow up on your question about the code that I asked you to delete from RValueBounds.   

The code that was deleted was handling an error case.  The error case was that the subexpression did not have bounds.  

The RValueBounds function does not need to check for errors in bounds.  That checking is done elsewhere and error messages will be generated if there is a problem.   Because error messages are generated elsewhere, no code will be generated.  The fact that RValueBounds may produce bounds that are invalid for incorrect input does not matter.   The RValueBounds function only needs to not crash if it is given IR for which an error message was generated.   

We specifically do not want to make the bounds depend on whether the subexpression bounds are correct or not.   Using the bounds of None in the case of an error involving the subexpression could cause confusing error messages for programmers.   The compiler might complain that it expected the entire dynamic_bounds_cast or assume_bounds_cast expression to have bounds.   That is a confusing error message for a programmer to see when the error involves only the subexpression.   

For this case,  the code in the class CheckBoundsDeclarations does the check that the subexpression has bounds as expected.  It generates an error message, if it does not.  This means that compilation will not succeed and no code will be generated.

The purpose of the RValueBounds function is to infer bounds for an expression.   Dynamic_bounds_cast and assume_bound_cast expression affect inference of bounds differently than C-style cast expressions.  For C-style cast expression, the bounds just pass through.  For dynamic_bounds_cast/assume_bounds_cast, these declare new bounds for the expression.

The following example illustrates the difference in the operators.  In the first case, a C-style cast (which is unnecessary) preserves the bounds of `a`, so it is OK for `b` to have bounds 10.  In the second case, a `dynamic_bounds_cast` declares new bounds for `a` that are smaller than the original bounds.  It is not OK for `c` to have bounds 10 because the bounds have been changed to be smaller by the `dynamic_bounds_cast`
```
int a checked[10];
array_ptr<int> b: count(10) = (array_ptr<int>) a;
array_ptr<int> c : count(10) = dynamic_bounds_cast<array_ptr<int>>(a, 5); // error
```